### PR TITLE
Add CUDA implementation of LFP computation

### PR DESCRIPTION
**Pull request by jafyvilla**
Using a basic, naive algorithm

### CODE REVIEW

**Comment by jafyvilla** [comment]
Retest this please

**Comment by jafyvilla** [comment]
Retest this please

**Comment by jafyvilla** [comment]
Retest this please

**Comment by jafyvilla** [comment]
Retest this please

**Comment by jafyvilla** [comment]
Retest this please

File: fivox/cuda/simpleLFP.cu [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    const float voxelPosY = yIndex * volInfo.voxelSize + volInfo.origin.y;
+    const float voxelPosZ = zIndex * volInfo.voxelSize + volInfo.origin.z;
+
+    // Compute directly the inverted value to gain performance in the for loop
+    const float cutOffDistance = 1.f / params.cutoff;
+
+    float current( 0.f );
+    for( unsigned int i = 0; i < params.numEvents; ++i )
+    {
+        const float value( values[i] );
```
Comment (hernando) [inline fivox/cuda/simpleLFP.cu]: Here there's an optimization opportunity. Instead of reading values[i] from each thread, you could read a a different value to shared memory (and the same for the event xyz coordinates). Even more interesting would be to compare the performance between using shared memory and warp shuffle operations.

File: fivox/cuda/simpleLFP.cu [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+        // If center of the voxel within the event radius, use the
+        // voltage at the surface of the compartment (at 'radius' distance)
+        const float radius( radii[i] );
+        // Comparison is inverted, as we are using the reciprocal values
+        // (radius is already stored like that from the loader)
+        if( length > radius )
+            current += value * radius; // mA
+        else
+            current += value * length; // mA
```
Comment (hernando) [inline fivox/cuda/simpleLFP.cu]: small optimization:

    current += value * min(radius, length);

File: fivox/volumeHandler.h [hunk 3]
```diff
@@ -47,10 +48,7 @@ class VolumeHandler
      * @param extent 3D vector containing the width, height and depth of the
      * volume in micrometers
      */
-    VolumeHandler( const size_t size, const vmml::Vector3f& extent )
-        : _size( size )
-        , _extent( extent )
-    {}
+    VolumeHandler( const size_t size, const vmml::Vector3f& extent );
```
Comment (hernando) [inline fivox/volumeHandler.h]: FIVOX_API declarations missing.

File: fivox/cudaImageSource.hxx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    gpuErrchk( cudaMemcpy( posX, source->getPositionsX(), fsize,
+                           cudaMemcpyHostToDevice ));
+    gpuErrchk( cudaMemcpy( posY, source->getPositionsY(), fsize,
+                           cudaMemcpyHostToDevice ));
+    gpuErrchk( cudaMemcpy( posZ, source->getPositionsZ(), fsize,
+                           cudaMemcpyHostToDevice ));
+    gpuErrchk( cudaMemcpy( radii, source->getRadii(), fsize,
+                           cudaMemcpyHostToDevice ));
+    gpuErrchk( cudaMemcpy( values, source->getValues(), fsize,
+                           cudaMemcpyHostToDevice ));
```
Comment (hernando) [inline fivox/cudaImageSource.hxx]: Here you have static and dynamic data. Is there a way to keep the static data between frames?

File: fivox/cuda/simpleLFP.cu [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        const float distanceX = voxelPosX - eventsX[i];
+        const float distanceY = voxelPosY - eventsY[i];
+        const float distanceZ = voxelPosZ - eventsZ[i];
+
+        const float distance2( distanceX * distanceX +
+                               distanceY * distanceY +
+                               distanceZ * distanceZ );
+
+        // Use the reciprocal of sqrt (faster in AVX assembly)
+        const float length = 1.f / sqrtf( distance2 );
```
Comment (hernando) [inline fivox/cuda/simpleLFP.cu]: Use an intrinsic and remove the comment about Intel processors.
http://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__SINGLE.html#group__CUDA__MATH__INTRINSIC__SINGLE

File: fivox/CMakeLists.txt [hunk 6]
```diff
@@ -42,13 +43,26 @@ set(FIVOX_SOURCES
...
 set(FIVOX_LINK_LIBRARIES
   PUBLIC ${ITK_LIBRARIES} Brion Lunchbox vmmlib
   PRIVATE ${Boost_LIBRARIES} Brain)
 
+if(CUDA_FOUND)
+  list(APPEND FIVOX_LINK_LIBRARIES ${CUDA_LIBRARIES})
+endif()
+
+option(FIVOX_USE_CUDA "Use CUDA" ON)
+add_subdirectory(cuda)
```
Comment (tribal-tec) [inline fivox/CMakeLists.txt]: shouldn't be in the if() below?

File: fivox/cuda/simpleLFP.cu [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                           float* __restrict__ output )
+{
+    // 1D grid of 1D blocks
+    const int threadId = blockIdx.x * blockDim.x + threadIdx.x;
+
+    const int xIndex = threadId % volInfo.dimensions.x;
+    const int yIndex = ( threadId / volInfo.dimensions.x )
+                       % volInfo.dimensions.y;
+    const int zIndex = (( threadId / volInfo.dimensions.x )
+                        / volInfo.dimensions.y ) % volInfo.dimensions.z;
```
Comment (hernando) [inline fivox/cuda/simpleLFP.cu]: It looks to me that using a 3D block and grid would be slightly faster because you don't need divisions and modulos.

File: fivox/cuda/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) BBP/EPFL 2014-2016
+#                        Jafet.VillafrancaDiaz@epfl.ch
+
+include(CMakeDependentOption)
+
+if(NOT FIVOX_USE_CUDA OR NOT CUDA_FOUND)
```
Comment (tribal-tec) [inline fivox/cuda/CMakeLists.txt]: opposite of the if() one level up

File: fivox/cuda/simpleLFP.cu (hunk 7 above)
Comment (hernando) [inline fivox/cuda/simpleLFP.cu]: You are not checking anywhere if this thread is out of bounds.

File: fivox/cuda/simpleLFP.cu [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+                           cudaMemcpyHostToDevice ));
+
+    const int numVoxels = volInfo.dimensions.x *
+                          volInfo.dimensions.y *
+                          volInfo.dimensions.z;
+
+    // 1D grid of 1D blocks
+    dim3 blockSize( 512, 1 );
+    dim3 gridSize( ( numVoxels + blockSize.x * blockSize.y - 1 ) /
+                   ( blockSize.x * blockSize.y ), 1 );
```
Comment (hernando) [inline fivox/cuda/simpleLFP.cu]: If you keep this 1D remove blockSize.y from this expression.

File: fivox/cuda/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  message(STATUS "Disable cuda, used ${FIVOX_USE_CUDA}, found ${CUDA_FOUND}")
+  return()
+endif()
+if(CUDA_VERSION_MAJOR VERSION_LESS 4)
+  message(WARNING "CUDA 4.0 or later needed, CUDA support disabled")
+  return()
+endif()
+
+add_definitions(-DUSE_CUDA)
```
Comment (tribal-tec) [inline fivox/cuda/CMakeLists.txt]: redundant to top-level CMakeLists.txt. I'd add it to the target with target_compile_definitions() below

File: fivox/cuda/simpleLFP.cu [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+    cudaEvent_t start, stop;
+    gpuErrchk( cudaEventCreate( &start ));
+    gpuErrchk( cudaEventCreate( &stop ));
```
Comment (hernando) [inline fivox/cuda/simpleLFP.cu]: Call to cudaEventDestroy missing.

File: fivox/cudaImageSource.h [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+#define FIVOX_CUDAIMAGESOURCE_H
+
+#include <fivox/imageSource.h>
+#include <fivox/types.h>
+
+namespace fivox
+{
+
+/**
+ * Image source iterating over all events and summing their values into the
```
Comment (tribal-tec) [inline fivox/cudaImageSource.h]: wrong!

File: fivox/cuda/simpleLFP.h [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+struct VolumeInfo
+{
+    uint3 dimensions;
+    float voxelSize;
+    float3 origin;
+};
+
+float simpleLFP( float* posX, float* posY, float* posZ,
+                 float* radii, float* values, const Parameters& parameters,
+                 const VolumeInfo& volInfo, float* output );
```
Comment (hernando) [inline fivox/cuda/simpleLFP.h]: documentation missing

File: fivox/cudaImageSource.hxx [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+template< typename TImage >
+void CudaImageSource< TImage >::GenerateData()
+{
+    auto image = Superclass::GetOutput();
+    image->Allocate();
+    image->FillBuffer( 0 );
+
+    auto source = Superclass::_eventSource;
+    float* posX;
```
Comment (tribal-tec) [inline fivox/cudaImageSource.hxx]: move closer to usage

File: fivox/cudaImageSource.hxx [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    gpuErrchk( cudaMemcpy( values, source->getValues(), fsize,
+                           cudaMemcpyHostToDevice ));
+
+    const int numVoxels = width * height * depth;
+
+    float* cudaOutput;
+    gpuErrchk( cudaMalloc( (void**)&cudaOutput, numVoxels * sizeof(float) ));
+    const float elapsed = cuda::simpleLFP( posX, posY, posZ, radii, values,
+                                           parameters, volInfo, cudaOutput );
+    LBINFO << "CUDA elapsed time: " << elapsed << " ms" << std::endl;
```
Comment (tribal-tec) [inline fivox/cudaImageSource.hxx]: Updating the progress bar is not easy to do I guess?

File: fivox/uriHandler.cpp [hunk 16]
```diff
@@ -495,11 +498,18 @@ ImageSourcePtr< TImage > URIHandler::newImageSource() const
         source = EventValueSummationImageSource< TImage >::New();
         break;
     default:
-        auto functorSource = FunctorImageSource< TImage >::New();
-        auto functor = newFunctor< TImage >();
-        functorSource->setFunctor( functor );
-        functor->setEventSource( eventSource );
-        source = functorSource;
+#ifdef USE_CUDA
```
Comment (tribal-tec) [inline fivox/uriHandler.cpp]: FIVOX_USE_CUDA?
On the cluster it means it will always use cuda. A runtime option to choose between cuda, CPU and maybe ispc could be needed depending on performance differences with different datasets and/or memory constrains.

File: fivox/CMakeLists.txt [hunk 17]
```diff
@@ -42,13 +43,26 @@ set(FIVOX_SOURCES
...
+  volumeHandler.cpp
   vsdLoader.cpp
 )
 
 set(FIVOX_LINK_LIBRARIES
   PUBLIC ${ITK_LIBRARIES} Brion Lunchbox vmmlib
   PRIVATE ${Boost_LIBRARIES} Brain)
 
+if(CUDA_FOUND)
+  list(APPEND FIVOX_LINK_LIBRARIES ${CUDA_LIBRARIES})
```
Comment (tribal-tec) [inline fivox/CMakeLists.txt]: should be done in the cuda folder below, plus it's PRIVATE

File: fivox/CMakeLists.txt (hunk 6 above)
Comment (hernando) [inline fivox/CMakeLists.txt]: +1

File: fivox/cuda/CMakeLists.txt (hunk 10 above)
Comment (hernando) [inline fivox/cuda/CMakeLists.txt]: I'm not sure that works with CUDA, in any case FIVOX_USE_CUDA from common_find_package is more adequate.

File: fivox/cuda/CMakeLists.txt [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+cuda_add_library(FivoxCuda SHARED ${CUDA_SOURCES})
+target_link_libraries(FivoxCuda Lunchbox)
+list(APPEND FIVOX_LINK_LIBRARIES FivoxCuda)
+set(FIVOX_LINK_LIBRARIES "${FIVOX_LINK_LIBRARIES}" PARENT_SCOPE)
+
+install(TARGETS FivoxCuda EXPORT ${PROJECT_NAME}Targets
+        LIBRARY DESTINATION lib COMPONENT COMMON)
+
+# Add a special target to clean nvcc generated files.
+cuda_build_clean_target()
```
Comment (chevtche) [inline fivox/cuda/CMakeLists.txt]: replace: 
    cuda_build_clean_target()
by:
    if(NOT TARGET clean_cuda_depends)
        cuda_build_clean_target()
    endif()

This fix the cmake step conflict with RTNeuron

File: fivox/cuda/simpleLFP.cu (hunk 7 above)
Comment (jafyvilla) [inline fivox/cuda/simpleLFP.cu]: I think that if I use a 3D block with a straightforward 1-to-1 mapping between threads and voxels, the Z dimension is very limited (64 in the Tesla K20m). I'm sure it can be improved, and it will, but that was not the goal of this PR anyway, but just to have a first, basic implementation.

The bounds check I accidentally removed, I will put it back.

File: fivox/cuda/simpleLFP.cu (hunk 1 above)
Comment (jafyvilla) [inline fivox/cuda/simpleLFP.cu]: I agree, but Grigori is working on that now, so I'd prefer to leave all the performance improvements for a different PR, and leave this version as a base for benchmarking.

File: fivox/cuda/simpleLFP.cu (hunk 9 above)
Comment (jafyvilla) [inline fivox/cuda/simpleLFP.cu]: I wanted to keep it generic, to be able to change the dimensions easily. We still need to experiment and compare with different sizes.

File: fivox/cuda/simpleLFP.cu (hunk 1 above)
Comment (hernando) [inline fivox/cuda/simpleLFP.cu]: It was just a comment.

File: fivox/cudaImageSource.hxx (hunk 15 above)
Comment (jafyvilla) [inline fivox/cudaImageSource.hxx]: Not trivial, it depends on the strategy we want to use (e.g. do it by chunks; start copying the output as it gets computed; etc). To be defined.

**Comment by jafyvilla** [comment]
Updated

File: fivox/CMakeLists.txt [hunk 19]
```diff
@@ -51,16 +51,13 @@ set(FIVOX_LINK_LIBRARIES
...
 
-if(CUDA_FOUND)
-  list(APPEND FIVOX_LINK_LIBRARIES ${CUDA_LIBRARIES})
-endif()
-
-option(FIVOX_USE_CUDA "Use CUDA" ON)
-add_subdirectory(cuda)
-if(FIVOX_USE_CUDA AND CUDA_FOUND)
-  add_definitions(-DUSE_CUDA)
+if(CUDA_FOUND AND FIVOX_USE_CUDA)
```
Comment (hernando) [inline fivox/CMakeLists.txt]: Are both needed after using common_find_package_disable?

File: fivox/CMakeLists.txt [hunk 20]
```diff
@@ -51,16 +51,13 @@ set(FIVOX_LINK_LIBRARIES
...
+  add_subdirectory(cuda)
+  list(APPEND PRIVATE FIVOX_LINK_LIBRARIES ${CUDA_LIBRARIES})
   list(APPEND FIVOX_PUBLIC_HEADERS cudaImageSource.h cudaImageSource.hxx)
   include_directories(SYSTEM ${CUDA_INCLUDE_DIRS})
+else()
+  message(STATUS "Disable cuda")
```
Comment (hernando) [inline fivox/CMakeLists.txt]: remove, it should be obvious from the configure output that Fivox is without CUDA. If it doesn't work when FIVOX_USE_CUDA is manually set to false, then common_find_package_disable needs fixing.

File: fivox/cuda/simpleLFP.h [hunk 21]
```diff
@@ -48,6 +48,24 @@ struct VolumeInfo
...
+ * distance
+ * @param volInfo structure containing the volume dimensions, the voxel size and
+ * the origin coordinates (bottom left corner)
+ * @param output array of the output values. Each value corresponds to a voxel,
+ * being its 3D indices flattened: first traversing along X, then Y
+ * (using "width" steps), and finally Z (using "width"*"height" planes)
+ * @return the time of execution, in milliseconds, of the computation in the
+ * CUDA kernel
+ */
 float simpleLFP( float* posX, float* posY, float* posZ,
```
Comment (hernando) [inline fivox/cuda/simpleLFP.h]: Add consts where possible.

**Comment by tribal-tec** [comment]
This still misses the runtime-check for CUDA to get a CUDA-enabled build on a non-gpu machine, right?

**Comment by jafyvilla** [comment]
Yes, I'll work on that as soon as I can

**Comment by jafyvilla** [comment]
@hernando thanks for the comments, but I'm afraid that I already hit the merge button while Github was malfunctioning, and the PR is already merged but the PR is still open :confused: . I guess I can close this one, and @chevtche can include your suggestions in the upcoming commit.
